src/main/scala/Subtyping.scala

```diff
@@ -0,0 +1,51 @@
+package fuselang
+
+import scala.math.{max,log10,ceil}
+import Syntax._
+import Errors.NoJoin
+
+/**
+ * Subtyping relations are only defined over the number hierarchy. Read 't1 < t2'
+ * as t1 is subtype of t2. The subtyping hierarchy is:
+ *
+ * sized(n) > sized(n - 1) ... > sized(1) > idx > static
+ *
+ * Note that all idx types and static types are equal to each other.
+ */
+object Subtyping {
+  def bitsNeeded(n: Int) = n match {
+    case 0 => 1
+    case n => ceil(log10(n)/log10(2)).toInt + 1
+  }
+
+  def areEqual(t1: Type, t2: Type) = (t1, t2) match {
+    case (_:TStaticInt, _:TStaticInt) => true
+    case (_:TIndex, _:TIndex) => true
+    case _ => t1 == t2
+  }
+  def isSubtype(sub: Type, sup: Type): Boolean = (sub, sup) match {
+    case (TSizedInt(v1), TSizedInt(v2)) => v1 <= v2
+    case (_:IntType, _:TSizedInt) => true
+    case (_:TStaticInt, _:TIndex) => true
+    case (TArray(tsub, _), TArray(tsup, _)) => isSubtype(tsup, tsub)
```

Comment:
Because arrays are mutable, it's probably safest to require invariant subtyping for them—i.e., either `areEqual(tsup, tsub)` or `isSubtype(tsup, tsub) && isSubtype(tsub, tsup)`. Otherwise, you could imagine something silly like this:

```
def foo(b: bit<1000>[20]) {
  b[0] := GIANT_NUMBER;
}

decl a : bit<10>[20];
foo(a);
```

i.e., to let an alias to the memory "disagree" about the right type of numbers the memory should contain.

Reply:
Fixed.